Add merge report UMAPs (#636)
AlexsLemonade/scpca-nf: templates/merge-report.Rmd

library(SingleCellExperiment)
library(ggplot2)

# Set default ggplot theme, customized for UMAPs
theme_set(

>   theme_bw() +

Reviewer: I think we don't want to use `theme_bw()` for the UMAPs because we want to remove the grid.

Author: I think we should keep it and just override the grid as I've done here. Either way, we'd have to override something - if we go with default ggplot, we'd have to remove the gray background, add in panel borders, and maybe also add facet strip outlines. I thought of maybe `theme_classic()` too, but we'd have to add in panel borders there too.. Either way something will have to get overridden, but `theme_bw()` gives us the panel borders we probably want to keep at least. Thoughts?

https://ggplot2.tidyverse.org/reference/ggtheme.html - 
![Screenshot 2024-01-02 at 12 29 38 PM](https://github.com/AlexsLemonade/scpca-nf/assets/4701111/39f3ca1c-1078-45e1-a290-bd9b557bd04e)

Reviewer: I think we want to keep the theme but then add this to when we actually create the UMAP so we remove the grid. 
https://github.com/AlexsLemonade/scpca-nf/blob/55e891648b801da5cb91384aa8fe5a951533e03e/templates/qc_report/celltypes_qc.rmd#L141-L144

Author: So, `panel.grid = element_blank(),` is already removing the grid. Here's where we are after https://github.com/AlexsLemonade/scpca-nf/pull/636/commits/f4ff899344a8f9eec33c6c937777bfa34f0691a4 (I used `firebrick3` which is just a color I like, open to discussion!)

![Screenshot 2024-01-02 at 1 18 15 PM](https://github.com/AlexsLemonade/scpca-nf/assets/4701111/5bb95864-81f9-46f2-9fed-27483ec6c6b7)

Author: noting this probably snuck past review since it was missing from the original image/PR! It was added here: https://github.com/AlexsLemonade/scpca-nf/pull/636/commits/16a46764e99fd680db553ed17b4be136702092a2

Reviewer: I like this without the grid, and I also am a fan of the color! I mean red is gross generally (for obvious sports reasons),  but it looks good here!

Author: >  I mean red is gross generally (for obvious sports reasons), but it looks good here!

😂 One reason I like this red since it isn't _as_ blinding as regular `red`, but this is also a good reason haha

```{r}
# To do: replace this with actual tables
# just a filler for now
table(params$merged_sce[[batch_column]])
```

## UMAPs

```{r, ref.label=c('integration-warning'), eval=TRUE, results='asis'}
```

> In this section, we present UMAPs of all libraries in the merged object, highlighting one library in each facet. 
> In each panel, cells from the given library are colored, while all other cells are in gray.

Reviewer: ```suggestion
In this section, we present UMAPs showing all cells from all libraries included in the merged object. 
For each library, a separate panel is shown, and cells from that library are colored, while all other cells are gray.
```
Just a little rewording that I think might clarify what is being shown.

> **These merged object UMAP coordinates differ from UMAPs in individual library files.**

Reviewer: I think if you're going to say this I would add a little more context. Maybe mention that after merging, we re-calculate PCA and UMAP. Because of this, the coordinates found in the merged object will differ from the coordinates found in the individual libraries.

Author: Added phrasing in https://github.com/AlexsLemonade/scpca-nf/pull/636/commits/3093c1f01e3e146de52b7943906bd784752d962b

The UMAPs presented here have been weighted so that each library contributes equally to the overall scaling.

```{r}
# Determine appropriate dimensions for UMAP based on number of libraries
n_libraries <- length(unique(params$merged_sce[[batch_column]]))

# Determine faceting layout based on n_libraries
facet_ncol <- dplyr::case_when(
  n_libraries <= 20 ~ 4,
  n_libraries <= 50 ~ 6,

>   n_libraries <= 80 ~ 8,

Reviewer: I think this is fine for now, but we might want to revisit when we actually create the merged object for projects this big. I'm worried things will get too small.

Author: > but we might want to revisit when we actually create the merged object for projects this big

Absolutely... I have no idea what's going to happen...

Author: Noting #641 has been opened

  .default = 10
)

# 3 inches per UMAP row?
umap_height <- ceiling(n_libraries / facet_ncol) * 3

# TODO: until we run the workflow through again after scpcaTools, this will not work. We will need to circle back to this.

> has_multiplex <- FALSE # temporary

Reviewer: You might be able to remove this, since I'm adding this in https://github.com/AlexsLemonade/scpca-nf/pull/632/files

Author: For now, maybe I'll keep a _temporary_ `is_multiplexed = FALSE` (with other `has_multiplex` -> `is_multiplexed`)?

Reviewer: Sure. I think we just want to wait to approve until #632 goes in then.

# Determine if library is multiplexed, for printing a warning.
# max_libraries_per_sample <- metadata(merged_sce)$sample_metadata |>
#  dplyr::count(library_id) |>
#  dplyr::pull(n) |>
#  max()

> # has_multiplex <- max_libraries_per_sample > 1

Reviewer: same applies to this, we probably wont' need it here since we should have `is_multiplexed` set when we create the tables.

```

```{r, eval = has_multiplex, results = 'asis'}
glue::glue("
 <div class=\"alert alert-info\">
 This merged object contains multiplexed data.
 Each UMAP facet displays a given library, which may be comprised of multiple samples.
  </div>
")
```

```{r, fig.width = 8, fig.height = umap_height}

> umap_df <- reducedDim(params$merged_sce, "UMAP") |>
>   as.data.frame() |>
>   tibble::rownames_to_column("library") |>
>   tidyr::separate(
>     library,
>     into = c("library", "barcode"),
>     sep = "-"
>   )

Reviewer: ```suggestion
umap_df <- reducedDim(params$merged_sce, "UMAP") |>
  as.data.frame() 
```
You won't need to do any of this since `library_id` is a column in the merged object.

Reviewer: Related to this, we will actually want to use the `batch_column` parameter.

Author: Yep, I realized about `batch_column` too....

But, the `reducedDims` don't have the `library_id` since that's in the `colData`... I might restore this code? Am I missing something?

Author: Yeah, I'm missing something. I'll need to _join_ the coldata to get the batch column

Reviewer: Oh yes you are right if you are just pulling the reducedDims and not the `colData`. I think we could grab the colData and merge with the reducedDims by the barcode/ colnames and then use the `params$batch_column` to make sure we are grabbing the correct ID's for the cells. Although the library ID's should be in the colnames, I think we want to maintain using the batch column.

Reviewer: I did not see your previous comment when writing this, but your plan sounds good!

ggplot(
  umap_df,

>   aes(x = UMAP1, y = UMAP2, color = library)

Reviewer: ```suggestion
  aes(x = UMAP1, y = UMAP2, color ={{params$batch_column}} )
```

) +
  # set points for all "other" points
  geom_point(
    data = dplyr::select(

>       umap_df, -library

Reviewer: ```suggestion
      umap_df, -{{params$batch_column}}
```
I actually am not sure this formatting will work, but I think it should 🤷‍♀️

    ),

>     color = "#ffcb05",

Reviewer: Just a reminder to sadly change these colors 😢

    alpha = 0.5,

>     size = 0.3

Reviewer: I think the points actually look kind of big and all lump together. I'm worried this will get worse as we have more libraries, so maybe going down to 0.1?
